### ISSUE TITLE
conda build 3.16.3 does not handle pin_compatible correctly for multiple outputs

### ISSUE THREAD

**Issue by bergtholdt** [conda/conda-build#3277]
### Actual Behavior

I get an error:

```
Could not get compatibility information for tbb package.  Is it one of your host dependencies?
```

### Expected Behavior

No errors since tbb is in the host section.

### Steps to Reproduce

I created a minimal recipe that you can run.

```yaml
package:
  name: test
  version: 1.0

build:
  number: 0

outputs:
  - name: libtest
    build:
      run_exports:
        - {{ pin_subpackage('libtest', exact=True) }}
    requirements:
      build:
        - {{ compiler('c') }}
      host:
        - tbb
        - tbb-devel
      run:
        - {{ pin_compatible('tbb') }}

  - name: test
    requirements:
      build:
        - {{ compiler('c') }}
      host:
        - python {{ python }}
        - {{ pin_subpackage('libtest', exact=True) }}

      run:
        - {{ pin_compatible('python') }}
        - {{ pin_subpackage('libtest', exact=True) }}
```

The error is actually happening during rendering for subpackage "test" not "libtest", however I observed the same behavior with swapped roles for test and libtest just with tbb replaced by python. The problem is, that pin_compatible is executed for all outputs, but should only be run for the current output. In jinja_context.py in function pin_compatible, I get versions like this:

```python
{'vs2015_runtime': ['14.15.26706', 'h3a45250_0'],
 'vc': ['14.1', 'h0510ff6_4'],
 'python': ['3.6.7', 'h33f27b4_1'],
 'certifi': ['2018.10.15', 'py36_0'],
 'wincertstore': ['0.2', 'py36h7fe50ca_0'],
 'setuptools': ['40.6.2', 'py36_0'],
 'wheel': ['0.32.3', 'py36_0'],
 'pip': ['18.1', 'py36_0'],
 'libtest': ['1.0']}
```

obviously tbb is not included since m.name() == 'test', we are here processing the test output. Still pin_compatible is run for tbb which results in no version info. But since m.extract_requirements_text() still returns

```yaml
  version: 1.0                                              
                                                            
build:                                                      
  number: 0                                                 
                                                            
outputs:                                                    
  - name: libtest                                           
    build:                                                  
      run_exports:                                          
        - {{ pin_subpackage('libtest', max_pin='x.x.x') }}  
    requirements:                                           
      build:                                                
        - {{ compiler('c') }}                               
      host:                                                 
        - tbb                                               
        - tbb-devel                                         
      run:                                                  
        - {{ pin_compatible('tbb') }}                       
```
the one for libtest. Actually it is matching the parent-package name: test (which it should not, but the test in get_recipe_text: is_output fails since test is also the parent's name). Anyways, 'pin_compatible' in m.extract_requirements_text() is True and thus the RuntimeError is raised.

Even when I rename the sub-package to something like py-test in the meta.yaml:
```yaml
package:
  name: test
  version: 1.0

build:
  number: 0

outputs:
  - name: libtest
    build:
      run_exports:
        - {{ pin_subpackage('libtest', exact=True) }}
    requirements:
      build:
        - {{ compiler('c') }}
      host:
        - tbb
        - tbb-devel
      run:
        - {{ pin_compatible('tbb') }}

  - name: test
    requirements:
      build:
        - {{ compiler('c') }}
      host:
        - python {{ python }}
        - {{ pin_subpackage('libtest', exact=True) }}

      run:
        - {{ pin_compatible('python') }}
        - {{ pin_subpackage('libtest', exact=True) }}
```
I still get the same error but now for the python package:

```
Could not get compatibility information for python package.  Is it one of your host dependencies?
```

I could not figure out how to correct this logic.

**Comment by msarahan** [comment]
You're running into the edge case of an output having the same name as the top-level output.  Conda-build's data model does not separate MetaData objects at each level, so it gets really hairy when there's both a top-level and output with the same name.

We recommend renaming your top-level recipe to something else.  It really doesn't matter what.  See our Numpy recipe for reference: https://github.com/AnacondaRecipes/numpy-feedstock/blob/master/recipe/meta.yaml#L5

**Comment by bergtholdt** [comment]
Yes,

I actually tested that case by using py-test as the name for the second output (I pasted the wrong contents in the 2nd meta.yaml, sorry) but that did not change things significantly (just the order that the packages got processed). I used this meta.yaml to test now with conda-build 3.17.1:

```yaml
package:
  name: test_lib_and_wrap
  version: 1.0

build:
  number: 0

outputs:
  - name: libtest
    build:
      run_exports:
        - {{ pin_subpackage('libtest', exact=True) }}
    requirements:
      build:
        - {{ compiler('c') }}
      host:
        - tbb
        - tbb-devel
      run:
        - {{ pin_compatible('tbb') }}

  - name: pytest
    requirements:
      build:
        - {{ compiler('c') }}
      host:
        - python {{ python }}
        - {{ pin_subpackage('libtest', exact=True) }}
      run:
        - {{ pin_compatible('python') }}
        - {{ pin_subpackage('libtest', exact=True) }}
```
I still get the error

    Could not get compatibility information for python package.  Is it one of your host dependencies?

during preparation for libtest. Since libtest does not have any dependency to python this should not happen. Somehow all pin_subpackage functions are processed for libtest, even though they only appear in other output sections. A possible fix is to not only check that pin_compatible exists in the m.extract_requirements_text() (jinja_context.py, ~line 259), but actually include the package name:

```python
    if (not compatibility and not permit_undefined_jinja and not bypass_env_check and
            "pin_compatible('{}'".format(package_name) in m.extract_requirements_text()):
 ```
Then it works. I still think that somehow in the processing hierarchy it should be tested if the pin_subpackage originates from another output than the current output and ignore the whole call, but this seems to work. I'll file a pull request.

**Comment by msarahan** [comment]
Thanks for digging into this more.  I agree with your evaluation of the situation, and look forward to your PR when you have time to submit it.

**Comment by bergtholdt** [comment]
In the end I used a regexp to have the check a little more generic: #3304

**Comment by msarahan** [comment]
fixed by #3304
